File: content/HMaA/lab3/src/main/java/ru/prokdo/fill/NumberGenerator.java
package ru.prokdo.fill;

import java.util.Random;

public final class NumberGenerator {
    private NumberGenerator() {
    }

    private final static Random RANDOM = new Random();

    public static Double generate(int min, int max, int precision) {
        Double integer = (double) RANDOM.nextInt(min, max + 1);

        if (precision == 0) return integer;

        Integer power = (int) Math.pow(10, precision);

        Integer fraction = RANDOM.nextInt(power);

        Double number = integer.doubleValue() + fraction.doubleValue() / power;

        return Math.round(number * power) / power.doubleValue();
    }

    public static Double generate(int max, int precision) {
        Double integer = (double) RANDOM.nextInt(max + 1);

        if (precision == 0) return integer;

        Integer power = (int) Math.pow(10, precision);

        Integer fraction = RANDOM.nextInt(power);

        Double number = integer.doubleValue() + fraction.doubleValue() / power;

        return Math.round(number * power) / power.doubleValue();
    }
}

File: content/HMaA/lab6/src/main/kotlin/ru/prokdo/model/util/random/NumberGenerator.kt
package ru.prokdo.model.util.random

import kotlin.math.pow
import kotlin.random.Random

object NumberGenerator {
    operator fun invoke(until: Int): Int = when {
        until < 0 -> -Random.nextInt(-until - 1)
        until > 0 -> Random.nextInt(until + 1)
        else -> 0
    }

    operator fun invoke(until: Int, precision: UInt): Double {
        var integer = when {
            until < 0 -> -Random.nextInt(-until - 1)
            until > 0 -> Random.nextInt(until + 1)
            else -> 0
        }

        val fraction = Random.nextInt(10.0.pow(precision.toInt()).toInt())

        return integer.toDouble() + fraction.toDouble() / 10.0.pow(precision.toInt())
    }

    operator fun invoke(from: Int, until: Int): Int = when {
        from < until -> Random.nextInt(from, until + 1)
        from > until -> Random.nextInt(until, from + 1)
        else -> from
    }

    operator fun invoke(from: Int, until: Int, precision: UInt): Double {
        var integer = when {
            from < until -> Random.nextInt(from, until + 1)
            from > until -> Random.nextInt(until, from + 1)
            else -> from
        }

        val fraction = Random.nextInt(10.0.pow(precision.toInt()).toInt())

        return integer.toDouble() + fraction.toDouble() / 10.0.pow(precision.toInt())
    }
}